spec/controllers/charges_controller_spec.rb

```diff
@@ -28,53 +29,195 @@
+        expect(payment_intent).to_not eq nil
+        expect(response.status).to eq 200
+      end
+    end
+
+    context 'when campaign has a project_id and a nonprofit with a fee_id' do
+      it 'should create a payment_intent that has the project_id, campaign_id and fee_id' do
+        @project = create(:project, id: project_id, square_location_id: square_location_id)
+        @fee = create(
+          :fee, 
+          id: fee_id,
+          active: true, 
+          multiplier: 0.1
+        )
+        @nonprofit = create(
+          :nonprofit,
+          id: nonprofit_id,
+          fee_id: @fee.id
+        )
+        @campaign = create(
+          :campaign,
+          id: campaign_id,
+          active: true,
+          seller_id: nil,
+          project_id: @project.id,
+          nonprofit_id: @nonprofit.id
+        )
+
+        payment_params = create_payment_params(campaign: @campaign, project: @project)
+
+        allow(SquareManager::PaymentCreator)
+          .to receive(:call)
+          .with(payment_params)
+          .and_return(mock_response)
+
+        cparams = create_charge_params(is_square: true, campaign: @campaign, is_distribution: false)
+        response = post :create, params: cparams, as: :json
+        payment_intent = PaymentIntent.find_by(
+                          campaign_id: campaign_id, 
+                          project_id: project_id, 
+                          fee_id: fee_id
+                         ) 
+        expect(payment_intent).to_not eq nil
+        expect(response.status).to eq 200
+      end
+    end
+  end
+
   ###############
   ### Helpers ###
   ###############
-  def charge_params(seller, is_square = false)
+
+  def create_charge_params(seller: nil, is_square: false, campaign: nil, is_distribution: true)
+    seller_id = seller.seller_id if seller
+    campaign_id = campaign.id if campaign
+    project_id = campaign.project_id if campaign
+    line_items = create_line_items(project_id: project_id)
     params = {
-      seller_id: seller.seller_id,
+      seller_id: seller_id,
       line_items: line_items,
       email: 'cthulhu@rlyeh.com',
       is_square: is_square,
       name: 'Cthulhu',
       idempotency_key: '42',
       is_subscribed: true,
-      is_distribution: true
+      is_distribution: is_distribution,
+      campaign_id: campaign_id,
+      project_id: project_id
     }
-
     params.merge! nonce: 'cnon:card-nonce-ok' if is_square
     params
   end
 
-  def line_items
-    [{
-      'amount' => 50,
-      'currency' => 'usd',
-      'quantity' => 1,
-      'item_type' => 'donation'
-    }]
+  def default_line_items 
+    { 
+      'amount' => 50, 
+      'currency' => 'usd', 
+      'quantity' => 1, 
+      'item_type' => 'donation' 
+    }
+  end
+
+  # TODO (billy-yuan): Update create_line_items so that transaction fees can be
+  # included as a separate line_items group
+  def create_line_items(extra_line_items = {})
```

Comment:
Examples of `extra_line_items` include `project_id` and `seller_id` but there's no typechecking or anything in place at the moment